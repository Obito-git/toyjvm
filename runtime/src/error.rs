use crate::ClassId;
use crate::class_loader::ClassLoaderErr;
use crate::rt::LinkageError;
use crate::rt::constant_pool::error::RuntimePoolError;
use common::jtype::HeapAddr;
use common::utils::cursor::CursorError;
use thiserror::Error;

//TODO: there is a way to not copy-paste macros? refactor later
#[macro_export]
macro_rules! throw_unsupported_operation {
    ($fmt:literal $(, $args:expr)* $(,)?) => {
        Err(JavaExceptionFromJvm::JavaLang(
            JavaLangError::UnsupportedOperationException(format!($fmt $(, $args)*)),
        ))
    };
    ($msg:expr) => {
        Err(JavaExceptionFromJvm::JavaLang(
            JavaLangError::UnsupportedOperationException($msg.into()),
        ))
    };
}

#[macro_export]
macro_rules! throw_arithmetic_exception {
    ($fmt:literal $(, $args:expr)* $(,)?) => {
        Err(JavaExceptionFromJvm::JavaLang(
            JavaLangError::ArithmeticException(format!($fmt $(, $args)*)),
        ))
    };
    ($msg:expr) => {
        Err(JavaExceptionFromJvm::JavaLang(
            JavaLangError::ArithmeticException($msg.into()),
        ))
    };
}

#[macro_export]
macro_rules! throw_negative_array_size_exception {
    ($fmt:literal $(, $args:expr)* $(,)?) => {
        Err(JavaExceptionFromJvm::JavaLang(
            JavaLangError::NegativeArraySizeException(format!($fmt $(, $args)*)),
        ))
    };
    ($msg:expr) => {
        Err(JavaExceptionFromJvm::JavaLang(
            JavaLangError::NegativeArraySizeException($msg.into()),
        ))
    };
}

#[macro_export]
// TODO: can have fixed message template
macro_rules! throw_index_out_of_bounds_exception {
    ($fmt:literal $(, $args:expr)* $(,)?) => {
        Err(JavaExceptionFromJvm::JavaLang(
            JavaLangError::ArrayIndexOutOfBoundsException(format!($fmt $(, $args)*)),
        ))
    };
    ($msg:expr) => {
        Err(JavaExceptionFromJvm::JavaLang(
            JavaLangError::ArrayIndexOutOfBoundsException($msg.into()),
        ))
    };
}

#[derive(Debug, Error)]
pub enum JvmError {
    #[error("LinkageError: {0}")]
    Linkage(#[from] LinkageError),
    #[error(transparent)]
    Cursor(#[from] CursorError),
    #[error("RuntimeConstantPoolError: {0}")]
    RuntimePool(#[from] RuntimePoolError),
    #[error(transparent)]
    ClassLoader(#[from] ClassLoaderErr),
    #[error("MissingAttributeInConstantPoll")]
    MissingAttributeInConstantPoll,
    #[error("ConstantNotFoundInRuntimePool")]
    ConstantNotFoundInRuntimePool,
    #[error("TrailingBytes")]
    TrailingBytes,
    #[error("ClassNotFoundException: {0}")]
    ClassNotFound(String),
    #[error("stack overflow")]
    StackOverflow,
    #[error("Frame stack is empty")]
    FrameStackIsEmpty,
    #[error("Operand stack is empty")]
    OperandStackIsEmpty,
    #[error("OutOfMemory")]
    OutOfMemory,
    #[error("Could not find or load main class {0}")]
    NoMainClassFound(String),
    #[error("NoSuchMethod: {0}")]
    NoSuchMethod(String),
    #[error("NoSuchField: {0}")]
    NoSuchFieldError(String),
    #[error("LocalVariableNotFound: {0}")]
    LocalVariableNotFound(u8),
    #[error("LocalVariableNotInitialized: {0}")]
    LocalVariableNotInitialized(u8),
    #[error("TypeDescriptorErr: {0}")]
    TypeDescriptorErr(#[from] common::TypeDescriptorErr),
    #[error("NullPointerException")]
    NullPointerException,
    #[error("InstructionErr: {0}")]
    InstructionErr(#[from] common::InstructionErr),
    #[error("ClassMirrorIsAlreadyCreated")]
    ClassMirrorIsAlreadyCreated,
    #[error("Method is not expecting to be abstract `{0}`")]
    MethodIsAbstract(String),
    #[error("UnexpectedType: `{0}`")]
    UnexpectedType(String),
    #[error("JavaExceptionThrown: `{0}`")]
    JavaExceptionThrown(HeapAddr),
    #[error("Uninitialized")]
    Uninitialized,
    #[error("WrongHeapAddress: `{0}`")]
    WrongHeapAddress(HeapAddr),
    #[error("TODO map to correct error: `{0}`")]
    Todo(String),
    #[error("JavaLangError: {0}")]
    JavaException(#[from] JavaExceptionFromJvm),
}

// TODO: everything below needs to be refactored
#[derive(Debug, Error)]
pub enum JavaExceptionFromJvm {
    #[error("LinkageError: {0}")]
    JavaLang(JavaLangError),
}

impl JavaExceptionFromJvm {
    pub fn as_reference(&self) -> JavaExceptionReference {
        match self {
            JavaExceptionFromJvm::JavaLang(err) => err.as_reference(),
        }
    }

    pub fn get_message(&self) -> &String {
        match self {
            JavaExceptionFromJvm::JavaLang(err) => err.get_message(),
        }
    }
}

pub struct JavaExceptionReference {
    pub class: &'static str,
    pub name: &'static str,
    pub descriptor: &'static str,
}

#[derive(Debug, Error)]
pub enum JavaLangError {
    #[error("java.lang.ArithmeticException: {0}")]
    ArithmeticException(String),
    #[error("java.lang.UnsupportedOperationException: {0}")]
    UnsupportedOperationException(String),
    #[error("java.lang.ArrayIndexOutOfBoundsException: {0}")]
    ArrayIndexOutOfBoundsException(String),
    #[error("java.lang.NegativeArraySizeException: {0}")]
    NegativeArraySizeException(String),
}

impl JavaLangError {
    pub fn as_reference(&self) -> JavaExceptionReference {
        match self {
            JavaLangError::ArithmeticException(_) => JavaExceptionReference {
                class: "java/lang/ArithmeticException",
                name: "<init>",
                descriptor: "(Ljava/lang/String;)V",
            },
            JavaLangError::UnsupportedOperationException(_) => JavaExceptionReference {
                class: "java/lang/UnsupportedOperationException",
                name: "<init>",
                descriptor: "(Ljava/lang/String;)V",
            },
            JavaLangError::ArrayIndexOutOfBoundsException(_) => JavaExceptionReference {
                class: "java/lang/ArrayIndexOutOfBoundsException",
                name: "<init>",
                descriptor: "(Ljava/lang/String;)V",
            },
            JavaLangError::NegativeArraySizeException(_) => JavaExceptionReference {
                class: "java/lang/NegativeArraySizeException",
                name: "<init>",
                descriptor: "(Ljava/lang/String;)V",
            },
        }
    }

    pub fn get_message(&self) -> &String {
        match self {
            JavaLangError::ArithmeticException(msg) => msg,
            JavaLangError::UnsupportedOperationException(msg) => msg,
            JavaLangError::ArrayIndexOutOfBoundsException(msg) => msg,
            JavaLangError::NegativeArraySizeException(msg) => msg,
        }
    }
}
