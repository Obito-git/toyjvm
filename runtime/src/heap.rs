// TODO: very primitive implementation, ok for right now

use crate::JvmError;
use crate::rt::class::class::Class;
use crate::rt::class::field::Field;
use crate::rt::constant_pool::reference::NameAndTypeReference;
use common::jtype::{HeapAddr, Value};
use std::sync::Arc;
use tracing_log::log::debug;

pub enum HeapObject {
    Instance(ClassInstance),
    String(String),
}

#[derive(Clone)]
pub struct ClassInstance {
    class: Arc<Class>,
    fields: Vec<Value>,
}

impl ClassInstance {
    pub fn get_field(&mut self, index: usize) -> &mut Value {
        self.fields.get_mut(index).expect("invalid field index")
    }

    pub fn class(&self) -> &Arc<Class> {
        &self.class
    }
}

/// https://docs.oracle.com/javase/specs/jvms/se24/html/jvms-2.html#jvms-2.5.3
pub struct Heap {
    objects: Vec<HeapObject>,
}

impl Heap {
    pub fn new() -> Self {
        debug!("Initializing Heap...");
        Self {
            objects: Vec::new(),
        }
    }

    fn push(&mut self, obj: HeapObject) -> HeapAddr {
        let idx = self.objects.len();
        self.objects.push(obj);
        idx
    }

    //TODO: how to handle super classes?
    pub fn alloc_instance(&mut self, class: Arc<Class>) -> HeapAddr {
        let fields = class
            .fields()
            .iter()
            .map(|field| field.descriptor().resolved().get_default_value())
            .collect();
        self.push(HeapObject::Instance(ClassInstance { class, fields }))
    }

    pub fn alloc_string<S: Into<String>>(&mut self, s: S) -> HeapAddr {
        self.push(HeapObject::String(s.into()))
    }

    pub fn get(&self, h: HeapAddr) -> &HeapObject {
        self.objects.get(h).expect("heap: invalid handle (get)")
    }

    pub fn get_instance(&mut self, h: &HeapAddr) -> &ClassInstance {
        let heap_obj = self.get(*h);
        match heap_obj {
            HeapObject::Instance(inst) => inst,
            _ => panic!("get_by_ref called with non-instance HeapObject",),
        }
    }

    pub fn get_instance_field(&mut self, h: &HeapAddr, nat: &NameAndTypeReference) -> &Value {
        let instance = self.get_instance(h);
        let slot = instance.class.get_field_index(nat).unwrap();
        instance.fields.get(slot).unwrap()
    }

    pub fn get_mut(&mut self, h: HeapAddr) -> &mut HeapObject {
        self.objects
            .get_mut(h)
            .expect("heap: invalid handle (get_mut)")
    }

    pub fn write_instance_field(
        &mut self,
        h: HeapAddr,
        field_nat: &NameAndTypeReference,
        val: Value,
    ) -> Result<(), JvmError> {
        match self.get_mut(h) {
            HeapObject::Instance(instance) => {
                let slot = instance.class.get_field_index(field_nat)?;
                instance.fields[slot] = val;
            }
            _ => panic!("heap: write_instance_field on non-instance"),
        }
        Ok(())
    }
}
