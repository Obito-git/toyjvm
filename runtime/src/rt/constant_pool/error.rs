use crate::rt::constant_pool::RuntimeConstantType;
use common::{MethodDescriptorErr, TypeDescriptorErr};
use thiserror::Error;

#[derive(Debug, Error)]
pub enum RuntimePoolError {
    #[error(transparent)]
    MethodDescriptor(#[from] MethodDescriptorErr),
    #[error(transparent)]
    TypeDescriptor(#[from] TypeDescriptorErr),
    #[error("WrongIndex")]
    WrongIndex(u16),
    #[error("TypeError")]
    TypeError(u16, RuntimeConstantType, RuntimeConstantType), //TODO: named args?
    #[error("TryingToAccessUnresolved")]
    TryingToAccessUnresolved(u16, RuntimeConstantType),
}
