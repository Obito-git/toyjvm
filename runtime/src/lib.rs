use crate::error::JvmError;
use crate::heap::Heap;
use crate::interpreter::Interpreter;
use crate::method_area::MethodArea;
use crate::native::NativeRegistry;
use crate::stack::FrameStack;
use std::cell::RefCell;
use std::rc::Rc;
use tracing_log::log::debug;

mod class_loader;
pub mod error;
mod heap;
mod interpreter;
mod method_area;
mod native;
pub mod rt;
pub mod stack;

pub type ClassId = usize;

#[derive(Debug)]
pub struct VmConfig {
    pub home: String,
    pub version: String,
    pub class_path: Vec<String>,
    pub initial_heap_size: usize,
    pub max_heap_size: usize,
    pub frame_stack_size: usize,
    pub operand_stack_size: usize,
}

//TODO: make it better
impl VmConfig {
    pub fn validate(&self) {
        if self.version != "24.0.2" {
            panic!(
                "Unsupported Java version: {}. Only 24.0.2 is supported.",
                self.version
            );
        }
    }
}

pub struct VirtualMachine {
    config: VmConfig,
    method_area: MethodArea,
    heap: Rc<RefCell<Heap>>,
    native_registry: NativeRegistry,
    frame_stack: FrameStack,
}

impl VirtualMachine {
    pub fn new(config: VmConfig) -> Result<Self, JvmError> {
        config.validate();
        let heap = Rc::new(RefCell::new(Heap::new()));
        let mut method_area = MethodArea::new(&config, heap.clone())?;
        debug!("Preloading java/lang/String...");
        let string_class = method_area.get_class("java/lang/String")?;
        heap.borrow().initialize(string_class);

        let native_registry = NativeRegistry::new();
        Ok(Self {
            frame_stack: FrameStack::new(&config),
            config,
            method_area,
            heap,
            native_registry,
        })
    }

    pub fn method_area(&mut self) -> &mut MethodArea {
        &mut self.method_area
    }

    pub fn heap(&self) -> Rc<RefCell<Heap>> {
        self.heap.clone()
    }
}

pub fn start(main_class: Vec<u8>, config: VmConfig) -> Result<(), JvmError> {
    debug!("Starting VM with config: {:?}", config);

    let vm = VirtualMachine::new(config)?;

    let mut interpreter = Interpreter::new(vm);
    interpreter.start(main_class)
}

#[cfg(test)]
impl serde::Serialize for VirtualMachine {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        use serde::ser::SerializeStruct;
        use std::ops::Deref;

        let mut state = serializer.serialize_struct("VirtualMachine", 3)?;
        state.serialize_field("heap", self.heap.borrow().deref())?;
        state.end()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::with_settings;
    use rstest::rstest;
    use std::fs;
    use std::path::{Path, PathBuf};

    const DISPLAY_SNAPSHOT_PATH: &str = "../snapshots";

    fn to_snapshot_name(path: &Path) -> String {
        let mut iter = path.iter().map(|s| s.to_string_lossy().to_string());
        for seg in iter.by_ref() {
            if seg == "test-classes" {
                break;
            }
        }
        let tail: Vec<String> = iter.collect();
        tail.join("-")
    }

    fn vm_config() -> VmConfig {
        let java_home = std::env::var("JAVA_HOME").expect("JAVA_HOME must be set for tests");

        VmConfig {
            home: java_home,
            version: "24.0.2".to_string(),
            class_path: vec![],
            initial_heap_size: 16 * 1024 * 1024,
            max_heap_size: 64 * 1024 * 1024,
            frame_stack_size: 256,
            operand_stack_size: 256,
        }
    }

    fn main_base_package_dir(path: &PathBuf) -> String {
        let mut dir = path.parent().expect("file should have a parent dir");

        dir = dir.parent().expect("should have parent dir");
        dir = dir.parent().expect("should have parent dir");

        dir.to_string_lossy().to_string()
    }

    #[rstest]
    #[trace]
    // Requires `testdata/compile-fixtures.sh` to be run to generate the .class files
    fn test_display(
        #[base_dir = "../target/test-classes/custom"]
        #[files("**/*Main.class")]
        path: PathBuf,
    ) {
        // Given
        let bytes = fs::read(&path).unwrap_or_else(|_| panic!("Can't read file {:?}", path));
        let mut base_config = vm_config();
        base_config.class_path.push(main_base_package_dir(&path));
        let vm = VirtualMachine::new(base_config).unwrap();
        let mut interpreter = Interpreter::new(vm);

        // When
        interpreter
            .start(bytes)
            .unwrap_or_else(|e| panic!("Interpreter failed for {:?} with error: {:?}", path, e));

        // Then
        with_settings!(
            {
                snapshot_path => DISPLAY_SNAPSHOT_PATH,
                prepend_module_to_snapshot => false,
            },
            {
                insta::assert_yaml_snapshot!(to_snapshot_name(&path), &interpreter);
            }
        );
    }
}
