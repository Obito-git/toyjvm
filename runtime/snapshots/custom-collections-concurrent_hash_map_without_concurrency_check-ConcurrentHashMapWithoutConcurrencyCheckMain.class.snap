---
source: runtime/src/lib.rs
expression: "&interpreter"
---
vm:
  heap:
    objects:
      - address: 0
        type: Instance
        value:
          class: java/lang/Runtime
          fields: []
      - address: 1
        type: Array
        class: java/io/ObjectStreamField
        elements:
          - Object: 2
          - Object: 6
          - Object: 15
      - address: 2
        type: Instance
        value:
          class: java/io/ObjectStreamField
          fields:
            - Object: 4
            - Object: ~
            - Object: 5
            - Object: ~
            - Integer: 0
            - Object: ~
            - Integer: 0
      - address: 3
        type: Array
        class: "[C"
        elements:
          - Integer: 115
          - Integer: 101
          - Integer: 103
          - Integer: 109
          - Integer: 101
          - Integer: 110
          - Integer: 116
          - Integer: 115
      - address: 4
        type: Instance
        value:
          class: java/lang/String
          fields:
            - Array: 3
            - Integer: 0
            - Integer: 0
            - Integer: 0
      - address: 5
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Integer: 0
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 6
        type: Instance
        value:
          class: java/io/ObjectStreamField
          fields:
            - Object: 8
            - Object: ~
            - Object: 13
            - Object: ~
            - Integer: 0
            - Object: ~
            - Integer: 0
      - address: 7
        type: Array
        class: "[C"
        elements:
          - Integer: 115
          - Integer: 101
          - Integer: 103
          - Integer: 109
          - Integer: 101
          - Integer: 110
          - Integer: 116
          - Integer: 77
          - Integer: 97
          - Integer: 115
          - Integer: 107
      - address: 8
        type: Instance
        value:
          class: java/lang/String
          fields:
            - Array: 7
            - Integer: 0
            - Integer: 0
            - Integer: 0
      - address: 9
        type: Array
        class: "[C"
        elements:
          - Integer: 105
          - Integer: 110
          - Integer: 116
      - address: 10
        type: Instance
        value:
          class: java/lang/String
          fields:
            - Array: 9
            - Integer: 0
            - Integer: 0
            - Integer: 0
      - address: 11
        type: Array
        class: java/lang/Class
        elements: []
      - address: 12
        type: Array
        class: java/io/ObjectStreamField
        elements: []
      - address: 13
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Integer: 0
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 14
        type: Array
        class: "[C"
        elements:
          - Integer: 48
          - Integer: 49
          - Integer: 50
          - Integer: 51
          - Integer: 52
          - Integer: 53
          - Integer: 54
          - Integer: 55
          - Integer: 56
          - Integer: 57
          - Integer: 97
          - Integer: 98
          - Integer: 99
          - Integer: 100
          - Integer: 101
          - Integer: 102
          - Integer: 103
          - Integer: 104
          - Integer: 105
          - Integer: 106
          - Integer: 107
          - Integer: 108
          - Integer: 109
          - Integer: 110
          - Integer: 111
          - Integer: 112
          - Integer: 113
          - Integer: 114
          - Integer: 115
          - Integer: 116
          - Integer: 117
          - Integer: 118
          - Integer: 119
          - Integer: 120
          - Integer: 121
          - Integer: 122
      - address: 15
        type: Instance
        value:
          class: java/io/ObjectStreamField
          fields:
            - Object: 17
            - Object: ~
            - Object: 13
            - Object: ~
            - Integer: 0
            - Object: ~
            - Integer: 0
      - address: 16
        type: Array
        class: "[C"
        elements:
          - Integer: 115
          - Integer: 101
          - Integer: 103
          - Integer: 109
          - Integer: 101
          - Integer: 110
          - Integer: 116
          - Integer: 83
          - Integer: 104
          - Integer: 105
          - Integer: 102
          - Integer: 116
      - address: 17
        type: Instance
        value:
          class: java/lang/String
          fields:
            - Array: 16
            - Integer: 0
            - Integer: 0
            - Integer: 0
      - address: 18
        type: Instance
        value:
          class: jdk/internal/misc/Unsafe
          fields: []
      - address: 19
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Integer: 0
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 20
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Integer: 0
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 21
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Integer: 0
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 22
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Integer: 0
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 23
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Integer: 0
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 24
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Integer: 0
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 25
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Integer: 0
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 26
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Integer: 0
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 27
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Integer: 0
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 28
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Integer: 0
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 29
        type: Array
        class: "[C"
        elements:
          - Integer: 115
          - Integer: 105
          - Integer: 122
          - Integer: 101
          - Integer: 67
          - Integer: 116
          - Integer: 108
      - address: 30
        type: Instance
        value:
          class: java/lang/String
          fields:
            - Array: 29
            - Integer: 0
            - Integer: 0
            - Integer: 0
      - address: 31
        type: Array
        class: "[C"
        elements:
          - Integer: 116
          - Integer: 114
          - Integer: 97
          - Integer: 110
          - Integer: 115
          - Integer: 102
          - Integer: 101
          - Integer: 114
          - Integer: 73
          - Integer: 110
          - Integer: 100
          - Integer: 101
          - Integer: 120
      - address: 32
        type: Instance
        value:
          class: java/lang/String
          fields:
            - Array: 31
            - Integer: 0
            - Integer: 0
            - Integer: 0
      - address: 33
        type: Array
        class: "[C"
        elements:
          - Integer: 98
          - Integer: 97
          - Integer: 115
          - Integer: 101
          - Integer: 67
          - Integer: 111
          - Integer: 117
          - Integer: 110
          - Integer: 116
      - address: 34
        type: Instance
        value:
          class: java/lang/String
          fields:
            - Array: 33
            - Integer: 0
            - Integer: 0
            - Integer: 0
      - address: 35
        type: Array
        class: "[C"
        elements:
          - Integer: 99
          - Integer: 101
          - Integer: 108
          - Integer: 108
          - Integer: 115
          - Integer: 66
          - Integer: 117
          - Integer: 115
          - Integer: 121
      - address: 36
        type: Instance
        value:
          class: java/lang/String
          fields:
            - Array: 35
            - Integer: 0
            - Integer: 0
            - Integer: 0
      - address: 37
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Integer: 0
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 38
        type: Array
        class: "[C"
        elements:
          - Integer: 118
          - Integer: 97
          - Integer: 108
          - Integer: 117
          - Integer: 101
      - address: 39
        type: Instance
        value:
          class: java/lang/String
          fields:
            - Array: 38
            - Integer: 0
            - Integer: 0
            - Integer: 0
      - address: 40
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Integer: 0
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 41
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Integer: 0
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 42
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Integer: 0
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 43
        type: Instance
        value:
          class: java/util/concurrent/ConcurrentHashMap
          fields:
            - Array: 54
            - Array: ~
            - Long: 1
            - Integer: 12
            - Integer: 0
            - Integer: 0
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 44
        type: Array
        class: "[C"
        elements:
          - Integer: 106
          - Integer: 97
          - Integer: 118
          - Integer: 97
          - Integer: 46
          - Integer: 118
          - Integer: 101
          - Integer: 114
          - Integer: 115
          - Integer: 105
          - Integer: 111
          - Integer: 110
      - address: 45
        type: Instance
        value:
          class: java/lang/String
          fields:
            - Array: 44
            - Integer: 0
            - Integer: 560567564
            - Integer: 0
      - address: 46
        type: Array
        class: "[C"
        elements:
          - Integer: 50
          - Integer: 52
      - address: 47
        type: Instance
        value:
          class: java/lang/String
          fields:
            - Array: 46
            - Integer: 0
            - Integer: 0
            - Integer: 0
      - address: 48
        type: Array
        class: java/io/ObjectStreamField
        elements: []
      - address: 49
        type: Instance
        value:
          class: java/lang/String$CaseInsensitiveComparator
          fields: []
      - address: 50
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Integer: 0
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 51
        type: Array
        class: "[C"
        elements:
          - Integer: 98
          - Integer: 121
          - Integer: 116
          - Integer: 101
      - address: 52
        type: Instance
        value:
          class: java/lang/String
          fields:
            - Array: 51
            - Integer: 0
            - Integer: 0
            - Integer: 0
      - address: 53
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Integer: 0
            - Object: ~
            - Array: ~
            - Object: ~
            - Object: ~
            - Object: ~
            - Object: ~
      - address: 54
        type: Array
        class: java/util/concurrent/ConcurrentHashMap$Node
        elements:
          - Object: ~
          - Object: ~
          - Object: ~
          - Object: ~
          - Object: ~
          - Object: 55
          - Object: ~
          - Object: ~
          - Object: ~
          - Object: ~
          - Object: ~
          - Object: ~
          - Object: ~
          - Object: ~
          - Object: ~
          - Object: ~
      - address: 55
        type: Instance
        value:
          class: java/util/concurrent/ConcurrentHashMap$Node
          fields:
            - Integer: 560575589
            - Object: 45
            - Object: 47
            - Object: ~
    string_pool:
      - string: "24"
        address: 47
      - string: baseCount
        address: 34
      - string: byte
        address: 52
      - string: cellsBusy
        address: 36
      - string: int
        address: 10
      - string: java.version
        address: 45
      - string: segmentMask
        address: 8
      - string: segmentShift
        address: 17
      - string: segments
        address: 4
      - string: sizeCtl
        address: 30
      - string: transferIndex
        address: 32
      - string: value
        address: 39
last_pop_frame:
  locals:
    - ~
    - Object: 43
    - Object: 47
  operands: []
  pc: 28
